lib/synapse/service_watcher/base.rb

```diff
@@ -95,13 +100,45 @@ def validate_discovery_opts
     end
 
     def set_backends(new_backends)
-      if @keep_default_servers
-        @backends = @default_servers + new_backends
+      # Aggregate and deduplicate all potential backend service instances.
+      new_backends = (new_backends + @default_servers) if @keep_default_servers
+      new_backends = new_backends.uniq {|b|
+        [b['host'], b['port'], b.fetch('name', '')]
+      }
+
+      if new_backends.to_set == @backends.to_set
+        return false
+      end
+
+      if new_backends.empty?
+        if @default_servers.empty?
+          if @use_previous_backends
+            # Discard this update
+            log.warn "synapse: no default servers for service #{@name};" \
```

Comment:
Should this (and the next `warn` statement) also mention that there are no backends for the service? E.g., "synapse: no backends for service #{@name} and no default servers configured; using previous backends: #{backends.inspect}"